File: DevLog/Infra/Common/FirestorePath.swift
//
//  FirestorePath.swift
//  DevLog
//
//  Created by opfic on 3/26/26.
//

enum FirestorePath {
    private enum Collection: String {
        case users
        case userData
        case counters
        case todoLists
        case activity
        case activityEvents
        case notifications
        case webPages
    }

    enum UserData: String {
        case info
        case tokens
        case settings
        case categories
    }

    enum Counter: String {
        case todo
    }

    static func user(_ uid: String) -> String {
        "\(Collection.users.rawValue)/\(uid)"
    }

    static func userData(_ uid: String, document: UserData? = nil) -> String {
        let path = "\(user(uid))/\(Collection.userData.rawValue)"
        guard let document else { return path }
        return "\(path)/\(document.rawValue)"
    }

    static func counter(_ uid: String, document: Counter) -> String {
        "\(user(uid))/\(Collection.counters.rawValue)/\(document.rawValue)"
    }

    static func todos(_ uid: String) -> String {
        "\(user(uid))/\(Collection.todoLists.rawValue)"
    }

    static func todo(_ uid: String, todoId: String) -> String {
        "\(todos(uid))/\(todoId)"
    }

    static func activity(_ uid: String) -> String {
        "\(user(uid))/\(Collection.activity.rawValue)"
    }

    static func activityDaily(_ uid: String, dayKey: String) -> String {
        "\(activity(uid))/\(dayKey)"
    }

    static func activityEvents(_ uid: String) -> String {
        "\(user(uid))/\(Collection.activityEvents.rawValue)"
    }

    static func activityEvent(_ uid: String, documentId: String) -> String {
        "\(activityEvents(uid))/\(documentId)"
    }

    static func notifications(_ uid: String) -> String {
        "\(user(uid))/\(Collection.notifications.rawValue)"
    }

    static func webPages(_ uid: String) -> String {
        "\(user(uid))/\(Collection.webPages.rawValue)"
    }

    static func webPage(_ uid: String, documentId: String) -> String {
        "\(webPages(uid))/\(documentId)"
    }
}

File: Firebase/functions/src/common/firestorePath.ts
export namespace FirestorePath {
    enum Collection {
        users = "users",
        userData = "userData",
        todoLists = "todoLists",
        activity = "activity",
        activityEvents = "activityEvents",
        notifications = "notifications",
        notificationDispatches = "notificationDispatches",
        webPages = "webPages"
    }

    export enum UserDataDocument {
        info = "info",
        tokens = "tokens",
        settings = "settings",
        categories = "categories"
    }

    export function user(userId: string): string {
        return `${Collection.users}/${userId}`;
    }

    export function userData(userId: string, document?: UserDataDocument): string {
        const path = `${user(userId)}/${Collection.userData}`;
        if (!document) { return path; }
        return `${path}/${document}`;
    }

    export function todos(userId: string): string {
        return `${user(userId)}/${Collection.todoLists}`;
    }

    export function todo(userId: string, todoId: string): string {
        return `${todos(userId)}/${todoId}`;
    }

    export function activity(userId: string): string {
        return `${user(userId)}/${Collection.activity}`;
    }

    export function activityDaily(userId: string, dayKey: string): string {
        return `${activity(userId)}/${dayKey}`;
    }

    export function activityEvents(userId: string): string {
        return `${user(userId)}/${Collection.activityEvents}`;
    }

    export function activityEvent(userId: string, eventId: string): string {
        return `${activityEvents(userId)}/${eventId}`;
    }

    export function notifications(userId: string): string {
        return `${user(userId)}/${Collection.notifications}`;
    }

    export function notification(userId: string, notificationId: string): string {
        return `${notifications(userId)}/${notificationId}`;
    }

    export function notificationDispatches(userId: string): string {
        return `${user(userId)}/${Collection.notificationDispatches}`;
    }

    export function notificationDispatch(userId: string, dispatchId: string): string {
        return `${notificationDispatches(userId)}/${dispatchId}`;
    }

    export function webPages(userId: string): string {
        return `${user(userId)}/${Collection.webPages}`;
    }

    export function webPage(userId: string, documentId: string): string {
        return `${webPages(userId)}/${documentId}`;
    }
}
